File: common/src/main/java/cc/turtl/chiselmon/api/filter/match/FilterMatcher.java
package cc.turtl.chiselmon.api.filter.match;

import cc.turtl.chiselmon.api.filter.FilterConditionParser;
import cc.turtl.chiselmon.api.filter.FiltersUserData;
import cc.turtl.chiselmon.api.filter.RuntimeFilter;
import cc.turtl.chiselmon.ChiselmonStorage;
import cc.turtl.chiselmon.api.storage.StorageScope;
import com.cobblemon.mod.common.pokemon.Pokemon;

import java.util.Comparator;
import java.util.List;
import java.util.Optional;
import java.util.function.Predicate;

public class FilterMatcher {
    private static List<RuntimeFilter> CACHE = null;

    public static FilterMatchResult match(Pokemon pokemon) {
        List<RuntimeFilter> filters = getFilters();

        List<RuntimeFilter> matches = filters.stream()
                .filter(f -> f.condition().test(pokemon))
                .toList();

        Optional<RuntimeFilter> primary = matches.stream()
                .max(Comparator.comparing(RuntimeFilter::priority));

        return new FilterMatchResult(pokemon, primary, matches);
    }

    public static void invalidateCache() {
        CACHE = null;
    }

    private static List<RuntimeFilter> getFilters() {
        if (CACHE == null) {
            CACHE = createRuntimeFilters();
        }
        return CACHE;
    }

    private static List<RuntimeFilter> createRuntimeFilters() {
        FiltersUserData data = ChiselmonStorage.FILTERS.get(StorageScope.global());
        data.migrateAll();

        return data.getAll().values().stream()
                .map(def -> {
                    Predicate<Pokemon> condition;
                    try {
                        condition = FilterConditionParser.parse(def.conditionString).toPredicate();
                    } catch (Exception e) {
                        condition = p -> false;
                    }
                    return new RuntimeFilter(def.id, def.displayName, def.rgb, def.priority, condition);
                })
                .sorted(Comparator.comparing(RuntimeFilter::priority).reversed())
                .toList();
    }
}

File: common/src/main/kotlin/cc/turtl/chiselmon/core/api/filter/match/FilterMatcher.kt
package cc.turtl.chiselmon.core.api.filter.match

import cc.turtl.chiselmon.client.ChiselmonStorage
import cc.turtl.chiselmon.core.ChiselmonConstants
import cc.turtl.chiselmon.core.api.filter.FilterConditionParser
import cc.turtl.chiselmon.core.api.filter.RuntimeFilter
import cc.turtl.chiselmon.core.api.storage.Scope
import com.cobblemon.mod.common.pokemon.Pokemon
import java.util.function.Predicate

object FilterMatcher {
    private var cache: List<RuntimeFilter>? = null

    @JvmStatic
    fun match(pokemon: Pokemon): FilterMatchResult {
        val filters = getFilters()
        val matches = filters.filter { it.condition.test(pokemon) }
        val primary = matches.maxByOrNull { it.priority }
        return FilterMatchResult(pokemon, primary, matches)
    }

    @JvmStatic
    fun invalidateCache() {
        cache = null
    }

    private fun getFilters(): List<RuntimeFilter> {
        return cache ?: createRuntimeFilters().also { cache = it }
    }

    private fun createRuntimeFilters(): List<RuntimeFilter> {
        val data = ChiselmonStorage.FILTERS[Scope.global()]
        data.migrateAll()

        return data.all.values
            .map { def ->
                val condition: Predicate<Pokemon> = try {
                    FilterConditionParser.parse(def.conditionString).toPredicate()
                } catch (e: Exception) {
                    ChiselmonConstants.LOGGER.warn(
                        "Filter '{}' has an invalid condition '{}': {}",
                        def.id,
                        def.conditionString,
                        e.message
                    )
                    Predicate { false }
                }
                RuntimeFilter(def.id, def.displayName, def.rgb, def.priority, condition)
            }
            .sortedByDescending { it.priority }
    }
}